horovod/common/controller.cc

```diff
@@ -451,6 +488,16 @@ Response Controller::ConstructResponse(std::string& name) {
     }
   }
 
+  // If there is at least one rank that requested Join, communicate tensor sizes
+  // in the response, because joined ranks don't have this info.
+  if (joined_size > 0 && message_type == Request::ALLREDUCE) {
```

Comment:
What about Allgather?

Reply:
My intention is to make this PR about AllReduce only, and add AllGather in the next PR.

Reply:
If we exclude Allgather and Broadcast from this commit, then we should to be careful about handling the case where the user submits an Allgather or Broadcast request during a join.  Ideally, we should throw an error saying "Allgather is not supported with Join at this time. Specify sparse_to_dense=True if using DistributedOptimizer".  Something like that.